### ISSUE TITLE
Support for Trio in Stream-Unzip

### ISSUE THREAD

**Issue by kulothunganug** [uktrade/stream-unzip#98]
We are developing a GUI application that heavily relies on the trio framework for asynchronous programming. However, we've encountered an issue where stream-unzip does not seem to work with trio out of the box. This is a important feature for our project, and integrating trio compatibility into stream-unzip would be very useful for our use case.

Is it possible to consider this as a feature request for future releases? It would be greatly appreciated. Thank you!

trio (async library): https://github.com/python-trio/trio
kivy (gui library): https://kivy.org/doc/stable/api-kivy.app.html#asynchronous-app

**Comment by kulothunganug** [comment]
```
Traceback (most recent call last):
  File "/home/kulothunganug/projects/rubem_app/test/t.py", line 47, in <module>
    trio.run(main)
  File "/home/kulothunganug/.cache/pypoetry/virtualenvs/rubem-app-YWRGg0wq-py3.12/lib/python3.12/site-packages/trio/_core/_run.py", line 2306, in run
    raise runner.main_task_outcome.error
  File "/home/kulothunganug/projects/rubem_app/test/t.py", line 44, in main
    await download_and_unzip(url, dest)
  File "/home/kulothunganug/projects/rubem_app/test/t.py", line 23, in download_and_unzip
    async for file_name, file_size, unzipped_chunks in async_stream_unzip(
  File "/home/kulothunganug/.cache/pypoetry/virtualenvs/rubem-app-YWRGg0wq-py3.12/lib/python3.12/site-packages/stream_unzip.py", line 553, in async_stream_unzip
    async for name, size, chunks in to_async_iterable(unzipped_chunks):
  File "/home/kulothunganug/.cache/pypoetry/virtualenvs/rubem-app-YWRGg0wq-py3.12/lib/python3.12/site-packages/stream_unzip.py", line 531, in to_async_iterable
    value = await loop.run_in_executor(None, *get_args())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: trio.run received unrecognized yield message <Future pending cb=[_chain_future.<locals>._call_check_cancel() at /usr/lib/python3.12/asyncio/futures.py:387]>. Are you trying to use a library written for some other framework like asyncio? That won't work without some kind of compatibility shim.
^CException ignored in: <module 'threading' from '/usr/lib/python3.12/threading.py'>
Traceback (most recent call last):
  File "/usr/lib/python3.12/threading.py", line 1594, in _shutdown
    atexit_call()
  File "/usr/lib/python3.12/concurrent/futures/thread.py", line 31, in _python_exit
    t.join()
  File "/usr/lib/python3.12/threading.py", line 1149, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.12/threading.py", line 1169, in _wait_for_tstate_lock
    if lock.acquire(block, timeout):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt:
```

Scroll to the right to see this `Are you trying to use a library written for some other framework like asyncio? That won't work without some kind of compatibility shim.`

This is a snippet from our codebase

```py
import httpx
import trio
from stream_unzip import async_stream_unzip


async def zipped_chunks(client, url):
    async with client.stream("GET", url) as r:
        async for chunk in r.aiter_bytes(chunk_size=65536):
            yield chunk


async def download_and_unzip(url, dest):
    async with httpx.AsyncClient() as client:
        total_size = 0
        head = await client.head(url)
        total_size = int(head.headers.get("Content-Length", 0))

        if total_size == 0:
            print("Unable to fetch file size. Download progress will not be shown.")

        downloaded_size = 0

        async for file_name, file_size, unzipped_chunks in async_stream_unzip(
            zipped_chunks(client, url),
            password=None,
        ):
            with open(file_name.decode("utf-8"), "wb") as f:
                async for chunk in unzipped_chunks:
                    f.write(chunk)
                    downloaded_size += len(chunk)

                    if total_size != 0:
                        percent_complete = (downloaded_size / total_size) * 100
                        print(f"Download Progress: {percent_complete:.2f}%")
                        print(f"Unzipping {file_name.decode('utf-8')}...")

    print("Download and unzip completed. Files extracted.")


async def main():
    url = "https://example.com/your-zip-file.zip"
    dest = "downloaded_zip"
    await download_and_unzip(url, dest)


trio.run(main)
```

It needs a zip url, but I hope you get the issue here

Thanks

**Comment by michalc** [comment]
Ah so it kinda looks like the issue comes from what you could call the async shim in stream-unzip: allowing its sync code to run from an async context.

So I guess there are 3-ish options: make the shim itself trio-compatible, or do away with the async and make a “pure” async version that is both asyncio and trio compatible, or (as the error suggests) but _another_ layer of shim in front of async_stream_unzip?

**Comment by michalc** [comment]
Opted for making the shim trio-compatible. Done in https://github.com/uktrade/stream-unzip/pull/100 and released in v0.0.94.

Will close this, but of course feel free to post an issue if there are problems with it
